File: Scala/compoundlab/src/main/scala/DuckCall.scala


class DuckCall(observable: Option[IQuackableObservable] = None) extends IQuackable {

	private val name = "Duck Call"
	private val _observable : IQuackableObservable = observable getOrElse new Observable(this)

	override def registerObserver(observer : IObserver) : Unit = {
		this._observable.registerObserver(observer)
	}

	override def notifyObservers: String = {
		this._observable.notifyObservers
	}

	override def quack : String = {
		notifyObservers
		"Duck called"
	}

	override def getName: String = {
		this.name
	}
}

File: Java/compoundlab/src/main/java/com/bbd/abstractFactoryPattern/DuckCall.java
package com.bbd.abstractFactoryPattern;


import java.util.ArrayList;
import java.util.List;
import java.util.Observer;

public class DuckCall implements IQuackable
{
	private String _name = "DuckCall";
	private IQuackObservable _observable;

	DuckCall()
	{
		this._observable = new Observable(this);
	}

	DuckCall(IQuackObservable observable)
	{
		this._observable = observable;
	}

	public String quack()
	{
		notifyObservers();
		return ("Duck calls");
	}

	public void registerObserver(IObserver observer)
	{
		this._observable.registerObserver(observer);
	}

	public String notifyObservers()
	{
		return (this._observable.notifyObservers());
	}

	@Override
	public String getName() { return  this._name;}
}
